glean/src/server_events.rs

```diff
@@ -0,0 +1,308 @@
+struct PingPayload {
+    client_info: ClientInfo,
+    ping_info: PingInfo,
+    metrics: Metrics,
+    events: Vec<GleanEvent>,
+}
+
+#[derive(Serialize, Deserialize)]
+struct LogEnvelope {
+    // MozLog compliant format. https://wiki.mozilla.org/Firefox/Services/Logging
+    #[serde(rename = "Type")]
+    log_type: String,
+    // MozLog compliant format. https://wiki.mozilla.org/Firefox/Services/Logging
+    #[serde(rename = "Fields")]
+    fields: Ping,
+}
+
+impl GleanEventsLogger {
+    fn create_client_info(&self) -> ClientInfo {
+        // Fields with default values are required in the Glean schema, but not used in server context
+        ClientInfo {
+            telemetry_sdk_build: "glean_parser v15.0.2.dev17+g81fec69a".to_owned(),
+            first_run_date: "Unknown".to_owned(),
+            os: "Unknown".to_owned(),
+            os_version: "Unknown".to_owned(),
+            architecture: "Unknown".to_owned(),
+            app_build: "Unknown".to_owned(),
+            app_display_version: self.app_display_version.clone(),
+            app_channel: self.app_channel.clone(),
+        }
+    }
+
+    fn create_ping(
+        &self,
+        document_type: &str,
+        config: &RequestInfo,
+        payload: &PingPayload,
+    ) -> Ping {
+        Ping {
+            document_namespace: self.app_id.clone(),
+            document_type: document_type.to_owned(),
+            document_version: "1".to_owned(),
+            document_id: Uuid::new_v4().to_string(),
+            user_agent: Some(config.user_agent.clone()),
+            ip_address: Some(config.ip_address.clone()),
+            payload: serde_json::to_string(payload).expect("unable to marshal payload to json."),
+        }
+    }
+
+    /// Method called by each ping-specific record method.
+    /// The goal is to construct the ping, wrap it in the envelope
+    /// and print to stdout.
+    fn record(
+        &self,
+        document_type: &str,
+        request_info: &RequestInfo,
+        metrics: Metrics,
+        events: Vec<GleanEvent>,
+    ) {
+        let telemetry_payload: PingPayload = PingPayload {
+            client_info: self.create_client_info(),
+            ping_info: PingInfo::default(),
+            metrics,
+            events,
+        };
+
+        let ping: Ping = self.create_ping(document_type, request_info, &telemetry_payload);
+
+        let envelope: LogEnvelope = LogEnvelope {
+            log_type: GLEAN_EVENT_MOZLOG_TYPE.to_owned(),
+            fields: ping,
+        };
+        let envelope_json =
+            serde_json::to_string(&envelope).expect("unable to marshal payload to json.");
+        println!("{}", envelope_json);
+    }
+}
+
+// Code below is generated based on the provided `metrics.yaml` file:
+
+// Metrics of the `event` type. Anything defined in `extra_keys` has it's own struct field.
+// The appended `Event` term to any metric of the event type implies the ping event.
+
+pub struct SyncstorageGetCollectionsEvent {
```

Comment:
Wondering if these should be in a different file. The stuff above this feel very much like generics, where this is more focused on the `syncstorage` product. 

I'm fine keeping them in place for now, though.